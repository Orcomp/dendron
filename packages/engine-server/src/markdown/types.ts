import {
  DendronConfig,
  DNoteRefLink,
  DVault,
  NotePropsV2,
} from "@dendronhq/common-all";
import { Parent } from "mdast";
import { Processor } from "unified";
import { LinkFilter } from "../topics/markdown/plugins/types";
import { DendronPubOpts } from "./remark/dendronPub";
import { WikiLinksOpts } from "./remark/wikiLinks";
export { VFile } from "vfile";

export type DendronASTNode = Parent & {
  notes?: NotePropsV2[];
  children?: Parent["children"] | DendronASTNode[];
};

export enum DendronASTTypes {
  WIKI_LINK = "wikiLink",
  REF_LINK = "refLink",
  PARAGRAPH = "paragraph",
}

export enum DendronASTDest {
  MD_ENHANCED_PREVIEW = "MD_ENHANCED_PREVIEW",
  MD_REGULAR = "MD_REGULAR",
  MD_DENDRON = "MD_DENDRON",
  HTML = "HTML",
}

export type DendronASTData = {
  dest: DendronASTDest;
  vault?: DVault;
  fname?: string;
  wikiLinkOpts?: WikiLinksOpts;
  config?: DendronConfig;
  overrides?: Partial<DendronPubOpts>;
  shouldApplyPublishRules?: boolean;
};

// NODES

export type WikiLinkNoteV4 = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  value: string;
  data: WikiLinkDataV4;
};

export type WikiLinkDataV4 = {
  alias: string;
  anchorHeader?: string;
  prefix?: string;
  vaultName?: string;
  filters?: LinkFilter[];
};

export type NoteRefNoteV4 = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  value: string;
  data: NoteRefDataV4;
};

export type NoteRefDataV4 = {
  link: DNoteRefLink;
};

export { Processor };
