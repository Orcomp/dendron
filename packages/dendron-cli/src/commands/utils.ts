import { createLogger, resolvePath } from "@dendronhq/common-server";
import { DEngineClientV2, EngineConnector } from "@dendronhq/engine-server";
import { LaunchEngineServerCommand } from "./launchEngineServer";
import _ from "lodash";
import yargs from "yargs";
const logger = createLogger();

export type SetupEngineCLIOpts = {
  wsRoot: string;
  enginePort?: number;
  init?: boolean;
};

export type SetupEngineResp = {
  wsRoot: string;
  engine: DEngineClientV2;
  port: number;
  server: any;
};

/**
 * Setup an engine based on CLI args
 */
export async function setupEngine(opts: SetupEngineCLIOpts): Promise<any> {
  let { wsRoot, enginePort, init } = _.defaults(opts, { init: true });
  let engine: DEngineClientV2;
  let port: number;
  let server: any;
  wsRoot = resolvePath(wsRoot, process.cwd());
  if (enginePort) {
    logger.info({
      ctx: "setupEngine",
      msg: "connecting to engine",
      enginePort,
    });
    const engineConnector = EngineConnector.getOrCreate({
      wsRoot,
    });
    await engineConnector.init({ portOverride: enginePort });
    engine = engineConnector.engine;
    port = enginePort;
    // dummy
    server = {
      close: () => {},
    };
  } else {
    logger.info({ ctx: "setupEngine", msg: "initialize new engine" });
    ({
      engine,
      port,
      server,
    } = await new LaunchEngineServerCommand().enrichArgs(opts));
    if (init) {
      await engine.init();
    }
  }
  return { wsRoot, engine, port, server };
}

/**
 * Add yargs based options to setup engine
 */
export function setupEngineArgs(args: yargs.Argv) {
  args.option("enginePort", {
    describe:
      "If set, connecto to running engine. If not set, create new instance of Dendron Engine",
  });
}
