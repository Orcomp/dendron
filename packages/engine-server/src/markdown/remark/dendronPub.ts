import { DendronError, NoteUtilsV2 } from "@dendronhq/common-all";
import _ from "lodash";
import { Image, Root } from "mdast";
import Unified, { Transformer } from "unified";
import { Node } from "unist";
import u from "unist-builder";
import visit from "unist-util-visit";
import { VFile } from "vfile";
import { findIndex, isNoteRefV2 } from "../../topics/markdown/plugins/inject";
import { SiteUtils } from "../../topics/site";
import { DendronASTDest, NoteRefDataV4, WikiLinkNoteV4 } from "../types";
import { MDUtilsV4 } from "../utils";
import { convertNoteRefAST, NoteRefsOpts } from "./noteRefs";
import { convertNoteRefASTV2 } from "./noteRefsV2";
import { addError, getNoteOrError } from "./utils";

type PluginOpts = NoteRefsOpts & {
  assetsPrefix?: string;
  insertTitle?: boolean;
  transformNoPublish?: boolean;
};

function plugin(this: Unified.Processor, opts?: PluginOpts): Transformer {
  const proc = this;
  const { dest, vault, fname, config, overrides } = MDUtilsV4.getDendronData(
    proc
  );
  function transformer(tree: Node, _file: VFile) {
    let root = tree as Root;
    const { error, engine } = MDUtilsV4.getEngineFromProc(proc);
    const insertTitle = !_.isUndefined(overrides?.insertTitle)
      ? overrides?.insertTitle
      : opts?.insertTitle;
    if (insertTitle && root.children) {
      if (!fname || !vault) {
        // TODO: tmp
        console.log(JSON.stringify(engine.notes));
        throw new DendronError({
          msg: `dendronPub - no fname or vault for node: ${JSON.stringify(
            tree
          )}`,
        });
      }
      const note = NoteUtilsV2.getNoteByFnameV4({
        fname,
        notes: engine.notes,
        vault: vault,
      });
      if (!note) {
        throw new DendronError({ msg: `no note found for ${fname}` });
      }
      const idx = _.findIndex(root.children, (ent) => ent.type !== "yaml");
      root.children.splice(
        idx,
        0,
        u("heading", { depth: 1 }, [u("text", note.title)])
      );
    }
    visit(tree, (node, _idx, parent) => {
      if (
        node.type === "wikiLink" &&
        dest !== DendronASTDest.MD_ENHANCED_PREVIEW
      ) {
        let _node = node as WikiLinkNoteV4;
        let value = node.value as string;
        // we change this later
        let valueOrig = value;
        let canPublish = true;
        const data = _node.data;
        if (error) {
          addError(proc, error);
        }

        const copts = opts?.wikiLinkOpts;
        if (opts?.transformNoPublish) {
          const notes = NoteUtilsV2.getNotesByFname({
            fname: valueOrig,
            notes: engine.notes,
            vault,
          });
          const { error, note } = getNoteOrError(notes, value);
          if (error) {
            value = "403";
            addError(proc, error);
          } else {
            if (!note || !config) {
              value = "403";
              addError(proc, new DendronError({ msg: "no note or config" }));
            } else {
              const vaults = engine.vaultsv3;
              const wsRoot = engine.wsRoot;
              canPublish = SiteUtils.canPublish({
                note,
                config: config.site,
                vaults,
                wsRoot,
              });
              if (!canPublish) {
                value = "403";
              }
            }
          }
        }
        if (copts?.useId && canPublish) {
          const notes = NoteUtilsV2.getNotesByFname({
            fname: valueOrig,
            notes: engine.notes,
            vault,
          });
          const { error, note } = getNoteOrError(notes, value);
          if (error) {
            addError(proc, error);
          } else {
            value = note!.id;
          }
        }
        const alias = data.alias ? data.alias : value;
        const href = `${copts?.prefix || ""}${value}.html${
          data.anchorHeader ? "#" + data.anchorHeader : ""
        }`;
        const exists = true;
        // for rehype
        //_node.value = newValue;
        //_node.value = alias;
        _node.data = {
          alias,
          permalink: href,
          exists: exists,
          hName: "a",
          hProperties: {
            // className: classNames,
            href,
          },
          hChildren: [
            {
              type: "text",
              value: alias,
            },
          ],
        };

        if (value === "403") {
          _node.data = {
            alias,
            hName: "a",
            hProperties: {
              "data-toggle": "popover",
              title: "This page has not yet sprouted",
              style: "cursor: pointer",
              "data-content": [
                `<a href="https://dendron.so/">Dendron</a> (the tool used to generate this site) lets authors selective publish content. You will see this page whenever you click on a link to an unpublished page`,
                "",
                "<img src='https://foundation-prod-assetspublic53c57cce-8cpvgjldwysl.s3-us-west-2.amazonaws.com/assets/images/not-sprouted.png'></img>",
              ].join("\n"),
            },
            hChildren: [
              {
                type: "text",
                value: alias,
              },
            ],
          };
        }
      }
      if (
        node.type === "refLink" &&
        dest !== DendronASTDest.MD_ENHANCED_PREVIEW
      ) {
        const ndata = node.data as NoteRefDataV4;
        const copts: NoteRefsOpts = {
          wikiLinkOpts: opts?.wikiLinkOpts,
          prettyRefs: opts?.prettyRefs,
        };
        const { data } = convertNoteRefAST({
          link: ndata.link,
          proc,
          compilerOpts: copts,
        });
        if (data) {
          parent!.children = data;
        }
      }
      if (
        node.type === "refLinkV2" &&
        dest !== DendronASTDest.MD_ENHANCED_PREVIEW
      ) {
        // we have custom compiler for markdown to handle note ref
        const ndata = node.data as NoteRefDataV4;
        const copts: NoteRefsOpts = {
          wikiLinkOpts: opts?.wikiLinkOpts,
          prettyRefs: opts?.prettyRefs,
        };
        const procOpts = proc.data("procOpts") as any;
        const { data } = convertNoteRefASTV2({
          link: ndata.link,
          proc,
          compilerOpts: copts,
          procOpts,
        });
        if (data) {
          if (parent!.children.length > 1) {
            const children = parent!.children;
            const idx = findIndex(children, isNoteRefV2);
            parent!.children = children
              .slice(0, idx)
              .concat(data)
              .concat(children.slice(idx + 1, -1));
          } else {
            parent!.children = data;
          }
        }
      }
      if (node.type === "image" && dest === DendronASTDest.HTML) {
        let imageNode = node as Image;
        if (opts?.assetsPrefix) {
          imageNode.url =
            "/" +
            _.trim(opts.assetsPrefix, "/") +
            "/" +
            _.trim(imageNode.url, "/");
        }
      }
    });
    return tree;
  }
  return transformer;
}

export { plugin as dendronPub };
export { PluginOpts as DendronPubOpts };
