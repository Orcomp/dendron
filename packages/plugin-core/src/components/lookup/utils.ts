import {
  DNodePropsQuickInputV2,
  DNodePropsV2,
  DNodeUtilsV2,
  DVault,
  NotePropsV2,
  NoteUtilsV2,
  VaultUtils,
} from "@dendronhq/common-all";
import { vault2Path } from "@dendronhq/common-server";
import _ from "lodash";
import path from "path";
import { Uri, ViewColumn, window } from "vscode";
import { Logger } from "../../logger";
import { VSCodeUtils } from "../../utils";
import { DendronWorkspace } from "../../workspace";
import { DendronBtn, getButtonCategory } from "./buttons";
import {
  CREATE_NEW_DETAIL,
  CREATE_NEW_LABEL,
  MORE_RESULTS_LABEL,
} from "./constants";
import { DendronQuickPickerV2 } from "./types";

export const UPDATET_SOURCE = {
  UPDATE_PICKER_FILTER: "UPDATE_PICKER_FILTER",
};

export function createNoActiveItem(vault: DVault): DNodePropsQuickInputV2 {
  const props = DNodeUtilsV2.create({
    fname: CREATE_NEW_LABEL,
    type: "note",
    vault,
  });
  return {
    ...props,
    label: CREATE_NEW_LABEL,
    detail: CREATE_NEW_DETAIL,
    alwaysShow: true,
  };
}
export function createMoreResults(): DNodePropsQuickInputV2 {
  // @ts-ignore
  return {
    label: MORE_RESULTS_LABEL,
    detail: "",
    alwaysShow: true,
  };
}

export function node2Uri(node: DNodePropsV2): Uri {
  const ext = node.type === "note" ? ".md" : ".yml";
  const nodePath = node.fname + ext;
  const wsRoot = DendronWorkspace.wsRoot();
  const vault = node.vault;
  const vpath = vault2Path({ wsRoot, vault });
  return Uri.file(path.join(vpath, nodePath));
}

export async function showDocAndHidePicker(
  uris: Uri[],
  picker: DendronQuickPickerV2
) {
  const ctx = "showDocAndHidePicker";
  const maybeSplitSelection = _.find(picker.buttons, (ent: DendronBtn) => {
    return getButtonCategory(ent) === "split" && ent.pressed;
  });
  let viewColumn = ViewColumn.Active;
  if (maybeSplitSelection) {
    const splitType = (maybeSplitSelection as DendronBtn).type;
    if (splitType === "horizontal") {
      viewColumn = ViewColumn.Beside;
    } else {
      // TODO: close current button
      // await commands.executeCommand("workbench.action.splitEditorDown");
    }
  }

  await Promise.all(
    uris.map(async (uri) => {
      return window.showTextDocument(uri, { viewColumn }).then(
        () => {
          Logger.info({ ctx, msg: "showTextDocument", fsPath: uri.fsPath });
          picker.hide();
          return;
        },
        (err) => {
          Logger.error({ ctx, err, msg: "exit" });
          throw err;
        }
      );
    })
  );
  return uris;
}

export type CreateQuickPickOpts = {
  title: string;
  placeholder: string;
  ignoreFocusOut?: boolean;
};

export class PickerUtilsV2 {
  static createDendronQuickPick(opts: CreateQuickPickOpts) {
    const { title, placeholder, ignoreFocusOut } = _.defaults(opts, {
      ignoreFocusOut: true,
    });
    const quickPick = window.createQuickPick<
      DNodePropsQuickInputV2
    >() as DendronQuickPickerV2;
    quickPick.title = title;
    quickPick.placeholder = placeholder;
    quickPick.ignoreFocusOut = ignoreFocusOut;
    quickPick.justActivated = true;
    quickPick.canSelectMany = false;
    quickPick.matchOnDescription = false;
    quickPick.matchOnDetail = false;
    return quickPick;
  }

  static createDendronQuickPickItem(
    opts: DNodePropsQuickInputV2
  ): DNodePropsQuickInputV2 {
    return {
      ...opts,
    };
  }

  static createDendronQuickPickItemFromNote(
    opts: NotePropsV2
  ): DNodePropsQuickInputV2 {
    return {
      ...opts,
      label: opts.fname,
    };
  }

  static dumpPicker(picker: DendronQuickPickerV2) {
    const activeItems = picker.activeItems.map((ent) =>
      NoteUtilsV2.toLogObj(ent)
    );
    const selectedItems = picker.selectedItems.map((ent) =>
      NoteUtilsV2.toLogObj(ent)
    );
    const value = picker.value;
    return { activeItems, selectedItems, value };
  }

  static getValue(picker: DendronQuickPickerV2) {
    return picker.value;
  }

  static getSelection(picker: DendronQuickPickerV2): DNodePropsQuickInputV2[] {
    return [...picker.selectedItems];
  }

  static filterCreateNewItem = (
    items: DNodePropsQuickInputV2[]
  ): DNodePropsQuickInputV2[] => {
    return _.reject(items, { label: CREATE_NEW_LABEL });
  };

  static filterDefaultItems = (
    items: DNodePropsQuickInputV2[]
  ): DNodePropsQuickInputV2[] => {
    return _.reject(
      items,
      (ent) =>
        ent.label === CREATE_NEW_LABEL || ent.label === MORE_RESULTS_LABEL
    );
  };

  /**
   * Reject all items that are over a given level
   * @param items
   * @param lvl
   */
  static filterByDepth = (
    items: DNodePropsQuickInputV2[],
    depth: number
  ): DNodePropsQuickInputV2[] => {
    return _.reject(items, (ent) => {
      return DNodeUtilsV2.getDepth(ent) > depth;
    });
  };

  static getFnameForOpenEditor(): string | undefined {
    const activeEditor = VSCodeUtils.getActiveTextEditor();
    if (activeEditor) {
      return path.basename(activeEditor.document.fileName, ".md");
    }
    return;
  }

  static getVaultForOpenEditor(opts?: { throwIfEmpty: boolean }): DVault {
    const vaults = DendronWorkspace.instance().vaultsv4;
    let vault: DVault;
    if (vaults.length > 1 && VSCodeUtils.getActiveTextEditor()?.document) {
      try {
        vault = VaultUtils.getVaultByNotePathV4({
          vaults,
          wsRoot: DendronWorkspace.wsRoot(),
          fsPath: VSCodeUtils.getActiveTextEditor()?.document.uri
            .fsPath as string,
        });
      } catch (err) {
        if (opts?.throwIfEmpty) {
          throw err;
        }
        return vaults[0];
      }
    } else {
      vault = vaults[0];
    }
    return vault;
  }

  static getOrPromptVaultForOpenEditor(): DVault {
    try {
      return PickerUtilsV2.getVaultForOpenEditor();
    } catch (err) {
      throw err;
      // TODO
    }
  }

  static isCreateNewNotePickForSingle(node: DNodePropsQuickInputV2): boolean {
    if (!node) {
      return true;
    }
    if (node.detail === CREATE_NEW_DETAIL || node.stub || node.schemaStub) {
      return true;
    } else {
      return false;
    }
  }

  static isCreateNewNotePick(node: DNodePropsQuickInputV2): boolean {
    if (!node) {
      return true;
    }
    if (node.detail === CREATE_NEW_DETAIL || node.stub || node.schemaStub) {
      return true;
    } else {
      return false;
    }
  }

  static isStringInputEmpty(value?: string) {
    if (_.isUndefined(value)) {
      return true;
    }
    if (_.isEmpty(value)) {
      return true;
    }
    return false;
  }

  static promptVault(): Promise<DVault | undefined> {
    const items = DendronWorkspace.instance().vaultsv4.map((ent) => ({
      ...ent,
      label: ent.fsPath,
    }));
    const resp = VSCodeUtils.showQuickPick(items) as Promise<
      DVault | undefined
    >;
    return resp;
  }

  static refreshButtons(opts: {
    quickpick: DendronQuickPickerV2;
    buttons: DendronBtn[];
    buttonsPrev: DendronBtn[];
  }) {
    opts.buttonsPrev = opts.quickpick.buttons.map((b: DendronBtn) => b.clone());
    opts.quickpick.buttons = opts.buttons;
  }

  static async refreshPickerBehavior(opts: {
    quickpick: DendronQuickPickerV2;
    buttons: DendronBtn[];
  }) {
    const resp = _.filter(opts.buttons, { pressed: true });
    return Promise.all(
      resp.map((bt) => {
        bt.handle({ quickPick: opts.quickpick });
      })
    );
  }

  static resetPaginationOpts(picker: DendronQuickPickerV2) {
    delete picker.moreResults;
    delete picker.offset;
    delete picker.allResults;
  }

  static slashToDot(ent: string) {
    return ent.replace(/\//g, ".");
  }
}
