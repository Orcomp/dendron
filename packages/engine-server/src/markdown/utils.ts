import {
  DendronConfig,
  DendronError,
  DEngineClientV2,
  DVault,
  getSlugger,
  NoteUtilsV2,
} from "@dendronhq/common-all";
// @ts-ignore
import mermaid from "@dendronhq/remark-mermaid";
// @ts-ignore
import rehypePrism from "@mapbox/rehype-prism";
import _ from "lodash";
import { Heading } from "mdast";
import { paragraph, root, text } from "mdast-builder";
import link from "rehype-autolink-headings";
// @ts-ignore
import katex from "rehype-katex";
import raw from "rehype-raw";
import slug from "rehype-slug";
import rehypeStringify from "rehype-stringify";
import remark from "remark";
import abbrPlugin from "remark-abbr";
import footnotes from "remark-footnotes";
import frontmatterPlugin from "remark-frontmatter";
import math from "remark-math";
import remarkParse from "remark-parse";
import remark2rehype from "remark-rehype";
import remarkStringify from "remark-stringify";
// @ts-ignore
import variables from "remark-variables";
import { default as unified, default as Unified, Processor } from "unified";
import { Node, Parent } from "unist";
import { backlinks } from "./remark/backlinks";
import { dendronPub, DendronPubOpts } from "./remark/dendronPub";
import { noteRefs, NoteRefsOpts } from "./remark/noteRefs";
import { noteRefsV2 } from "./remark/noteRefsV2";
import { wikiLinks, WikiLinksOpts } from "./remark/wikiLinks";
import { DendronASTData, DendronASTDest } from "./types";

const toString = require("mdast-util-to-string");

type ProcOpts = {
  engine: DEngineClientV2;
};

export class MDUtilsV4 {
  static findIndex(array: Node[], fn: any) {
    for (var i = 0; i < array.length; i++) {
      if (fn(array[i], i)) {
        return i;
      }
    }
    return -1;
  }

  static genMDMsg(msg: string): Parent {
    return root(paragraph(text(msg)));
  }

  static getDendronData(proc: Processor) {
    return proc.data("dendron") as DendronASTData;
  }

  static getFM(proc: Processor) {
    return proc.data("fm") as any;
  }

  static setDendronData(proc: Processor, data: Partial<DendronASTData>) {
    const _data = proc.data("dendron") as DendronASTData;
    return proc.data("dendron", { ..._data, ...data });
  }

  static getEngineFromProc(proc: Unified.Processor) {
    const engine = proc.data("engine") as DEngineClientV2;
    let error: DendronError | undefined;
    if (_.isUndefined(engine) || _.isNull(engine)) {
      error = new DendronError({ msg: "engine not defined" });
    }
    return {
      error,
      engine,
    };
  }

  static getNoteRefLvl(proc: Unified.Processor): number {
    return (proc.data("noteRefLvl") as number) || 0;
  }

  static setNoteRefLvl(proc: Unified.Processor, lvl: number) {
    return proc.data("noteRefLvl", lvl);
  }

  // @deprecate
  static isHeading(node: Node, text: string, depth?: number) {
    if (node.type !== "heading") {
      return false;
    }

    // wildcard is always true
    if (text === "*") {
      return true;
    }
    if (text) {
      var headingText = toString(node);
      return text.trim().toLowerCase() === headingText.trim().toLowerCase();
    }

    if (depth) {
      return (node as Heading).depth <= depth;
    }

    return true;
  }

  static matchHeading(
    node: Node,
    text: string,
    opts: { depth?: number; slugger: ReturnType<typeof getSlugger> }
  ) {
    const { depth, slugger } = opts;
    if (node.type !== "heading") {
      return false;
    }

    // wildcard is always true
    if (text === "*") {
      return true;
    }

    if (text) {
      var headingText = toString(node);
      return text.trim().toLowerCase() === slugger.slug(headingText.trim());
    }

    if (depth) {
      return (node as Heading).depth <= depth;
    }

    return true;
  }

  static remark() {
    let _proc = remark()
      .use(remarkParse, { gfm: true })
      .use(frontmatterPlugin, ["yaml"])
      .use({ settings: { listItemIndent: "1", fences: true, bullet: "-" } });
    return _proc;
  }

  static proc(opts: ProcOpts) {
    const { engine } = opts;
    const errors: DendronError[] = [];
    let _proc = remark()
      .use(remarkParse, { gfm: true })
      //.use(remarkStringify)
      .data("errors", errors)
      .data("engine", engine)
      .use(frontmatterPlugin, ["yaml"])
      .use({ settings: { listItemIndent: "1", fences: true, bullet: "-" } });
    return _proc;
  }

  static procFull(
    opts: ProcOpts & {
      dest: DendronASTDest;
      shouldApplyPublishRules?: boolean;
      vault?: DVault;
      config?: DendronConfig;
      fname?: string;
      wikiLinksOpts?: WikiLinksOpts;
      noteRefOpts?: NoteRefsOpts;
      publishOpts?: DendronPubOpts;
      mathOpts?: {
        katex?: boolean;
      };
      mermaid?: boolean;
    }
  ) {
    const { dest, vault, fname, config, shouldApplyPublishRules } = opts;
    let proc = this.proc(opts);
    if (vault && fname) {
      const engine = MDUtilsV4.getEngineFromProc(proc).engine;
      const note = NoteUtilsV2.getNoteByFnameV4({
        fname,
        notes: engine.notes,
        vault,
      });
      const fm = {
        ...note?.custom,
        title: note?.title,
      };
      proc = proc.data("fm", fm);
    }
    proc = proc
      .data("dendron", {
        dest,
        vault,
        fname,
        config,
        shouldApplyPublishRules,
      } as DendronASTData)
      //.use(extract, { name: "fm" })
      .use(abbrPlugin)
      .use(variables)
      .use(footnotes)
      .use(wikiLinks, opts.wikiLinksOpts)
      .use(backlinks)
      .use(noteRefsV2, {
        ...opts.noteRefOpts,
        wikiLinkOpts: opts.wikiLinksOpts,
      })
      .use(noteRefs, { ...opts.noteRefOpts, wikiLinkOpts: opts.wikiLinksOpts });
    if (opts.mathOpts?.katex) {
      proc = proc.use(math);
    }
    if (opts.mermaid) {
      proc = proc.use(mermaid, { simple: true });
    }
    // MD_DENDRON, convert back to itself, no need for transformations
    if (dest !== DendronASTDest.MD_DENDRON) {
      proc = proc.use(dendronPub, {
        ...opts.publishOpts,
        wikiLinkOpts: opts.wikiLinksOpts,
      });
    }
    proc = proc.data("procFull", proc().freeze());
    return proc;
  }

  static procRemark(opts: { proc?: Processor }) {
    const { proc } = opts;
    let _proc = proc || this.remark();
    return _proc.use(remarkParse, { gfm: true }).use(remarkStringify);
  }

  static procRehype(opts: {
    proc?: Processor;
    mdPlugins?: Processor[];
    mathjax?: boolean;
  }) {
    const { proc, mdPlugins } = _.defaults(opts, { mdPlugins: [] });
    let _proc = proc || unified().use(remarkParse, { gfm: true });
    mdPlugins.forEach((p) => {
      _proc = _proc.use(p);
    });
    _proc = _proc
      .use(remark2rehype, { allowDangerousHtml: true })
      .use(rehypePrism, { ignoreMissing: true })
      .use(raw)
      .use(slug)
      .use(link, {
        properties: {
          "aria-hidden": "true",
          class: "anchor-heading",
        },
        content: {
          type: "element",
          tagName: "svg",
          properties: {
            "aria-hidden": "true",
            viewBox: "0 0 16 16",
          },
          children: [
            {
              type: "element",
              tagName: "use",
              properties: {
                "xlink:href": "#svg-link",
              },
            },
          ],
        },
      });
    if (opts.mathjax) {
      _proc = _proc.use(katex);
    }
    return _proc.use(rehypeStringify);
  }
}
