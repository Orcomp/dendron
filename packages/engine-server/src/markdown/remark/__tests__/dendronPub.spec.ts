import { DEngineClientV2 } from "@dendronhq/common-all";
import {
  AssertUtils,
  ENGINE_HOOKS,
  NoteTestUtilsV4,
  runEngineTestV4,
} from "@dendronhq/common-test-utils";
import _ from "lodash";
import { DConfig } from "../../../config";
import { createEngine } from "../../../enginev2";
import { DendronASTData, DendronASTDest } from "../../types";
import { MDUtilsV4 } from "../../utils";
import { DendronPubOpts } from "../dendronPub";

function proc(
  engine: DEngineClientV2,
  dendron: DendronASTData,
  opts?: DendronPubOpts
) {
  return MDUtilsV4.procFull({ engine, ...dendron, publishOpts: opts });
}

describe("basics", () => {
  let engine: any;
  let dendronData: DendronASTData = {
    dest: DendronASTDest.HTML,
  };

  test("imagePrefix", () => {
    const out = proc(engine, dendronData, {
      assetsPrefix: "bond/",
    }).processSync(`![alt-text](image-url.jpg)`);
    expect(_.trim(out.toString())).toEqual("![alt-text](/bond/image-url.jpg)");
  });

  test("imagePrefix2", () => {
    const out = proc(engine, dendronData, {
      assetsPrefix: "/bond/",
    }).processSync(`![alt-text](/image-url.jpg)`);
    expect(_.trim(out.toString())).toEqual("![alt-text](/bond/image-url.jpg)");
  });

  describe("no publish", () => {
    test("basic", async () => {
      await runEngineTestV4(
        async ({ engine, vaults }) => {
          const vault = vaults[0];
          const config = DConfig.genDefaultConfig();
          config.site = {
            siteHierarchies: ["foo"],
            siteRootDir: "foo",
          };
          const resp = await MDUtilsV4.procRehype({
            proc: proc(
              engine,
              { ...dendronData, config, vault },
              {
                wikiLinkOpts: { useId: true },
                transformNoPublish: true,
              }
            ),
          }).process(`[[an alias|bar]]`);
          expect(resp).toMatchSnapshot();
          expect(
            await AssertUtils.assertInString({
              body: resp.contents as string,
              match: ["a data-toggle="],
            })
          ).toBeTruthy();
        },
        { expect, createEngine, preSetupHook: ENGINE_HOOKS.setupBasic }
      );
    });

    test("inside note ref, wikilink", async () => {
      await runEngineTestV4(
        async ({ engine, vaults }) => {
          const vault = vaults[0];
          const config = DConfig.genDefaultConfig();
          config.site = {
            siteHierarchies: ["foo"],
            siteRootDir: "foo",
          };
          const resp = await MDUtilsV4.procRehype({
            proc: proc(
              engine,
              {
                ...dendronData,
                config,
                vault,
                fname: "gamma",
                shouldApplyPublishRules: true,
              },
              {
                wikiLinkOpts: { useId: true },
                transformNoPublish: true,
              }
            ),
          }).process("[[alpha]]");
          expect(resp).toMatchSnapshot();
          expect(
            await AssertUtils.assertInString({
              body: resp.contents as string,
              match: ["This page has not yet sprouted"],
            })
          ).toBeTruthy();
        },
        {
          expect,
          createEngine,
          preSetupHook: async (opts) => {
            await ENGINE_HOOKS.setupLinks(opts);
            await NoteTestUtilsV4.createNote({
              fname: "gamma",
              body: `![[alpha]]`,
              vault: opts.vaults[0],
              wsRoot: opts.wsRoot,
            });
          },
        }
      );
    });

    test("inside note ref, note ref link", async () => {
      await runEngineTestV4(
        async ({ engine, vaults }) => {
          const vault = vaults[0];
          const config = DConfig.genDefaultConfig();
          config.site = {
            siteHierarchies: ["foo"],
            siteRootDir: "foo",
          };
          const resp = await MDUtilsV4.procRehype({
            proc: proc(
              engine,
              {
                ...dendronData,
                config,
                vault,
                fname: "gamma",
                shouldApplyPublishRules: true,
              },
              {
                wikiLinkOpts: { useId: true },
                transformNoPublish: true,
              }
            ),
          }).process("![[alpha]]");
          expect(resp).toMatchSnapshot();
          expect(resp.contents as string).toEqual("<p></p><p></p><p></p>");
        },
        {
          expect,
          createEngine,
          preSetupHook: async (opts) => {
            await ENGINE_HOOKS.setupLinks(opts);
            await NoteTestUtilsV4.createNote({
              fname: "gamma",
              body: `![[alpha]]`,
              vault: opts.vaults[0],
              wsRoot: opts.wsRoot,
            });
          },
        }
      );
    });
  });
});
