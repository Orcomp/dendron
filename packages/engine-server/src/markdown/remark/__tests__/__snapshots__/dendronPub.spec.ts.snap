// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basics no publish basic 1`] = `
VFile {
  "contents": "<p><a data-toggle=\\"popover\\" title=\\"This page has not yet sprouted\\" style=\\"cursor: pointer\\" data-content=\\"<a href=&#x22;https://dendron.so/&#x22;>Dendron</a> (the tool used to generate this site) lets authors selective publish content. You will see this page whenever you click on a link to an unpublished page

<img src=&#x27;https://foundation-prod-assetspublic53c57cce-8cpvgjldwysl.s3-us-west-2.amazonaws.com/assets/images/not-sprouted.png&#x27;></img>\\">an alias</a></p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`basics no publish inside note ref, note ref link 1`] = `
VFile {
  "contents": "<p></p><p></p><p></p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`basics no publish inside note ref, wikilink 1`] = `
VFile {
  "contents": "<p><a data-toggle=\\"popover\\" title=\\"This page has not yet sprouted\\" style=\\"cursor: pointer\\" data-content=\\"<a href=&#x22;https://dendron.so/&#x22;>Dendron</a> (the tool used to generate this site) lets authors selective publish content. You will see this page whenever you click on a link to an unpublished page

<img src=&#x27;https://foundation-prod-assetspublic53c57cce-8cpvgjldwysl.s3-us-west-2.amazonaws.com/assets/images/not-sprouted.png&#x27;></img>\\">alpha</a></p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
